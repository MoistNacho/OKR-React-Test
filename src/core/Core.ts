import { observable } from "mobx";

import Dialog from "./Dialog";
import PopUp from "./PopUp";
import Router from "./Router";

export default class Core {
  public dialog: Dialog;

  public router: Router;

  public popUp: PopUp;

  @observable
  public locale = "ko";

  private stores: {
    [key: string]: any;
  } = {};

  constructor() {
    this.dialog = new Dialog();
    this.router = new Router();
    this.popUp = new PopUp();

    // expose core during tests
    if ((window as any).Cypress) {
      (window as any).core = this;
    }
  }

  public addStore = (key: string, store: any) => {
    this.stores[key] = store;
  };

  public getStore = (key: string) => this.stores[key];

  public clearStores = () => {
    this.stores = {};
  };

  public reset = () => this.clearStores();
}
