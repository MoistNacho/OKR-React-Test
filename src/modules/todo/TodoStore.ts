import { action, observable } from "mobx";

export type TodoItem = {
  id: number;
  name: string;
};

export default class TodoStore {
  @observable
  public todoList: TodoItem[] = [
    { id: 0, name: "Javascript" },
    { id: 1, name: "React" },
    { id: 2, name: "Next.js" },
  ];

  @observable
  public nextId = 3;

  @action.bound
  public handleAddItem(name: string) {
    this.todoList = [...this.todoList, { id: this.nextId, name }];
    this.nextId += 1;
  }

  @action.bound
  public handleUpdateItem(id: number, name: string) {
    const newArray = this.todoList.map((item) => {
      if (item.id === id) {
        return { ...item, name };
      }

      return { ...item };
    });

    this.todoList = newArray;
  }

  @action.bound
  public handleRemoveItem(id: number) {
    const newArray = this.todoList.filter((item) => {
      if (item.id === id) return;

      return { ...item };
    });

    this.todoList = newArray;
  }
}
