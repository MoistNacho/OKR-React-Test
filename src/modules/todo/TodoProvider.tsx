import React from "react";

import { generateUseContext } from "lib/context";

import TodoStore from "./TodoStore";

export interface TodoContext {
  todoStore: TodoStore;
}

interface Props {
  children: React.ReactNode;
  stores: TodoContext;
}

const LocalContext = React.createContext<TodoContext | null>(null);

export const useTodoStore = generateUseContext(LocalContext);

const TodoProvider: React.FC<Props> = ({ children, stores }) => {
  return (
    <LocalContext.Provider value={stores}>{children}</LocalContext.Provider>
  );
};

export default TodoProvider;
