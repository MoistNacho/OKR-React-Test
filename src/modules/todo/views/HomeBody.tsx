import { observer } from "mobx-react";
import React from "react";
import styled from "styled-components";

import { useTodoStore } from "../TodoProvider";

import TodoAddForm from "./components/TodoAddForm";
import TodoList from "./components/TodoList";

const HomeBody = observer(() => {
  const { todoStore } = useTodoStore();
  const { handleAddItem } = todoStore;

  return (
    <HomeBodyWrap>
      <Title>Simple Todo List</Title>
      <TodoAddForm onCreate={handleAddItem} />
      <TodoList />
    </HomeBodyWrap>
  );
});

export default HomeBody;

const HomeBodyWrap = styled.div`
  width: 800px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  align-items: center;
  color: #303540;

  @media screen and (max-width: 800px) {
    width: 100%;
    padding: 0 10px;
  }
`;

const Title = styled.h2`
  margin: 12px auto;
`;
