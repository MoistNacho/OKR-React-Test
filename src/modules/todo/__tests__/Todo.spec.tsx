/* eslint-disable prefer-destructuring */
import { BaseTheme } from "@meshkorea/vroong-design-system-web";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import React from "react";
import renderer from "react-test-renderer";
import { ThemeProvider } from "styled-components";

import { CoreProvider } from "core";

import TodoProvider, { TodoContext } from "../TodoProvider";
import TodoStore from "../TodoStore";
import HomeBody from "../views/HomeBody";

describe("<TodoList />", () => {
  let stores: TodoContext;
  let wrraper: React.ReactElement;

  beforeEach(() => {
    stores = { todoStore: new TodoStore() };
    wrraper = (
      <CoreProvider>
        <ThemeProvider theme={BaseTheme}>
          <TodoProvider stores={stores}>
            <HomeBody />
          </TodoProvider>
        </ThemeProvider>
      </CoreProvider>
    );
  });

  it("renders with snapshot", () => {
    const component = renderer.create(wrraper);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("TodoList Integration Test", () => {
    let input: HTMLInputElement;
    let addButton: HTMLButtonElement;
    let editButton: HTMLElement;
    let updateButton: HTMLButtonElement;
    let removeButton: HTMLElement;

    beforeEach(() => {
      render(wrraper);
    });

    it("Item Add Test", async () => {
      input = screen.getByPlaceholderText("값을 입력해주세요");
      addButton = screen.getByRole("button", { name: "확인" });

      await userEvent.type(input, "New Item");
      await userEvent.click(addButton);
      const itemCount = screen.getByText("Item Count: 4");

      expect(itemCount).toBeInTheDocument();
      expect(stores.todoStore.todoList.length).toBe(4);
      expect(stores.todoStore.todoList[3].name).toBe("New Item");
    });

    it("Item Update Test", async () => {
      editButton = screen.getAllByRole("button", { name: "수정" })[0];

      await userEvent.click(editButton);
      input = screen.getByPlaceholderText("아이템 이름을 입력해주세요");
      updateButton = screen.getByRole("button", { name: "완료" });

      await userEvent.clear(input);
      await userEvent.type(input, "Update Item");
      await userEvent.click(updateButton);

      const changedItem = screen.getByText("1. Update Item");

      expect(changedItem).toBeInTheDocument();
      expect(stores.todoStore.todoList[0].name).toBe("Update Item");
    });

    it("Item Remove Test", async () => {
      removeButton = screen.getAllByRole("button", { name: "삭제" })[0];

      await userEvent.click(removeButton);
      const itemCount = screen.getByText("Item Count: 2");

      expect(itemCount).toBeInTheDocument();
      expect(stores.todoStore.todoList.length).toBe(2);
    });
  });
});
