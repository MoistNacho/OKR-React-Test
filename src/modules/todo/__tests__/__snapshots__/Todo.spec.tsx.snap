// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeBody /> renders with snapshot 1`] = `
<div
  className="sc-hCgVqe kzJvFz"
>
  <h2
    className="sc-eELIPF gvtDLV"
  >
    Simple Todo List
  </h2>
  <div
    className="sc-bPPhlf enuzHx"
  >
    <div
      className="sc-dZeWys eRorbB"
      width="400px"
    >
      <div
        className="sc-hbyLVd eJEjZG"
        disabled={false}
      >
        <input
          className="sc-oclUV dpdYNp"
          onChange={[Function]}
          onKeyDown={[Function]}
          onKeyPress={[Function]}
          placeholder="값을 입력해주세요"
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      className="sc-bczRLJ iehTtZ"
      disabled={true}
      onClick={[Function]}
      onKeyPress={[Function]}
      type="button"
    >
      확인
    </button>
  </div>
  <div
    className="sc-foDcoF iOljnP"
  >
    <div
      className="sc-hQXzsD IhLXJ"
    >
      Item Count: 
      3
    </div>
    <ul
      className="sc-eZMymg glaGss"
    >
      <li
        className="sc-hrZiYQ kqSlcB"
      >
        <div
          className="sc-caiQmJ ekyrCo"
        >
          <span>
            1. Javascript
          </span>
          <div
            className="sc-jnCuNW AxYum"
          >
            <button
              className="sc-bczRLJ iehTtZ"
              disabled={false}
              onClick={[Function]}
              onKeyPress={[Function]}
              type="button"
            >
              수정
            </button>
            <button
              className="sc-bczRLJ iehTtZ"
              disabled={false}
              onClick={[Function]}
              onKeyPress={[Function]}
              style={
                Object {
                  "color": "#fa5c5c",
                }
              }
              type="button"
            >
              삭제
            </button>
          </div>
        </div>
      </li>
      <li
        className="sc-hrZiYQ kqSlcB"
      >
        <div
          className="sc-caiQmJ ekyrCo"
        >
          <span>
            2. React
          </span>
          <div
            className="sc-jnCuNW AxYum"
          >
            <button
              className="sc-bczRLJ iehTtZ"
              disabled={false}
              onClick={[Function]}
              onKeyPress={[Function]}
              type="button"
            >
              수정
            </button>
            <button
              className="sc-bczRLJ iehTtZ"
              disabled={false}
              onClick={[Function]}
              onKeyPress={[Function]}
              style={
                Object {
                  "color": "#fa5c5c",
                }
              }
              type="button"
            >
              삭제
            </button>
          </div>
        </div>
      </li>
      <li
        className="sc-hrZiYQ kqSlcB"
      >
        <div
          className="sc-caiQmJ ekyrCo"
        >
          <span>
            3. Next.js
          </span>
          <div
            className="sc-jnCuNW AxYum"
          >
            <button
              className="sc-bczRLJ iehTtZ"
              disabled={false}
              onClick={[Function]}
              onKeyPress={[Function]}
              type="button"
            >
              수정
            </button>
            <button
              className="sc-bczRLJ iehTtZ"
              disabled={false}
              onClick={[Function]}
              onKeyPress={[Function]}
              style={
                Object {
                  "color": "#fa5c5c",
                }
              }
              type="button"
            >
              삭제
            </button>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
